internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,163 @@ import (
 )
 
+
+
+// Runtime2 is the VM.
+// Review: the object that implements this outlives actor calls.
+type Runtime2 struct {
+	CurerntEpoch() Epoch
+	// Review: why is this taking an offset? verify size of offset
+	Randomness(epoch Epoch, offset uint64) Randomness
+	// Review: the spec still has a TODO on this
+	// Review: not sure why the original has a messagereceipt instead of an InvocOutput
+	Send(input InvocInput) InvocOutput
+	// Review: how does this work? What qualifies as an "error"?
+	// Review: how is this any different than the caller just ignoring a returned error?
+	// SendAllowingErrors(input msg.InvocInput) msg.MessageReceipt
+
+	// Review: we might want a global Storage() Storage here
+}
+
+// InvocationContext is passed to the actors on each method call.
+// Review: this type + Runtime2 is mostly what the spec defines as Runtime.
+type InvocationContext interface {
+	// Review: grouped the VM level things out of the invocation itself
+	Runtime() Runtime2
+	Caller() address.Address
+	// Review: (is in spec, dont think we need it) idem to Caller().Equal(addr) :/
+	// ValidateCallerIs(caller addr.Address)
+	// Review: not sure what all this pattern matching adds up to, the "singleton" pattern does an equality check
+	// ValidateCallerMatches(CallerPattern)
+	// Review: the spec handles state and state changes a bit differently
+	// Review: our Storage is a bit of a mix of AcquireState and the Ipld{Get-Put}
+	// AcquireState() ActorStateHandle
+
+	// Review: The spec suggest handling return construction by the runtime, interesting but not strictly necessary
+	// SuccessReturn()     msg.InvocOutput
+	// ValueReturn(Bytes)  msg.InvocOutput
+	// ErrorReturn(exitCode exitcode.ExitCode) msg.InvocOutput
+
+	// Review: how does this work? is the spec expecting us to Panic? or just terminate the message execution?
+	// Review: this should be `Abort(string) -> msg.InvocOutput` if we follow the pattern
+	// Abort(string)
+
+	// Review: this is a bit trickier than abort in Go. This is a conditional early exit on the function which we cannot do.
+	// Review: `if <cond> { return AssertFailure("blah") }` is a bit closer to go if we are to try to follow this pattern.
+	// Assert(bool)
+
+	// Review: we didnt have this before, why does the actor need to know the value supplied?
+	ValueSupplied() types.AttoFIL
+
+	// Review: if this is just for the `init` actor then it should not be here.
+	// Review: the init actor is a special system actor, its ok for it to have a different interface with a bit more power.
+    // CreateActor(
+    //     cid                actor.StateCID
+    //     a                  addr.Address
+    //     constructorParams  actor.MethodParams
+    // )
+
+	// Review: we provide this through the Storage()
+    // IpldGet(c ipld.CID) union {Bytes, error}
+	// IpldPut(x ipld.Object) ipld.CID
+	// Review: is the expectation fo storage to be global or actor specific? this has security implications down the line
```

Comment:
Could you elaborate on the security implications?  At first glance I think it should not matter with hash links everywhere.

Reply:
Splitting global/actor-local storage means that the hashspace the actor has access to through the actor storage is restricted to objects he put in there in the past. Other actors cannot access those objects even if they have the CID (unless the CID is also present in their storage). 

It also allows us in the future to establish object provenance and storage consumption per actor.

Reply:
This is a tough question -- actor-specific storage is desirable for the reasons you mention, but on the other hand, we would like it to be the case that actor A can send a large object to actor B just by CID reference. If this also requires the runtime to inspect the contents of the message and copy all constituent CIDs to actor B's address space, that could get pretty messy (not to mention causing a lot of duplication for common data objects). @zixuanzh @jbenet Thoughts on this question?

Reply:
Getting both the isolation and the cheap "send" (with just the CID) is totally doable, the question is whether or not the spec really wants/needs either/both of those characteristics to achieve its goals.

Reply:
My sense is that the isolation property isn't critical at the moment, but others should feel free to weigh in.